aws/resource_aws_sagemaker_notebook_instance.go

```diff
@@ -0,0 +1,390 @@
+		NotebookInstanceName: aws.String(d.Id()),
+	}
+	notebook, err := conn.DescribeNotebookInstance(describeNotebookInput)
+	if err != nil {
+		if sagemakerErr, ok := err.(awserr.Error); ok && sagemakerErr.Code() == "ResourceNotFound" {
+			return nil
+		}
+		return fmt.Errorf("unable to find sagemaker notebook instance to delete %q: %s", d.Id(), err)
+	}
+	if *notebook.NotebookInstanceStatus != sagemaker.NotebookInstanceStatusFailed {
+		if err := stopSagemakerNotebookInstance(conn, d.Id()); err != nil {
+			return err
+		}
+	}
+
+	deleteOpts := &sagemaker.DeleteNotebookInstanceInput{
+		NotebookInstanceName: aws.String(d.Id()),
+	}
+
+	if _, err := conn.DeleteNotebookInstance(deleteOpts); err != nil {
+		return fmt.Errorf("error trying to delete sagemaker notebook instance %q: %s", d.Id(), err)
+	}
+
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{
+			sagemaker.NotebookInstanceStatusDeleting,
+		},
+		Target:  []string{""},
+		Refresh: sagemakerNotebookInstanceStateRefreshFunc(conn, d.Id()),
+		Timeout: 10 * time.Minute,
+	}
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("error waiting for sagemaker notebook instance %q to delete: %s", d.Id(), err)
+	}
+
+	return nil
+}
+
+func stopSagemakerNotebookInstance(conn *sagemaker.SageMaker, id string) error {
+	describeNotebookInput := &sagemaker.DescribeNotebookInstanceInput{
+		NotebookInstanceName: aws.String(id),
+	}
+	notebook, err := conn.DescribeNotebookInstance(describeNotebookInput)
+	if err != nil {
+		if sagemakerErr, ok := err.(awserr.Error); ok && sagemakerErr.Code() == "ResourceNotFound" {
+			return nil
+		}
+		return fmt.Errorf("unable to find sagemaker notebook instance %q: %s", id, err)
+	}
+	if *notebook.NotebookInstanceStatus == sagemaker.NotebookInstanceStatusStopped {
+		return nil
+	}
+
+	stopOpts := &sagemaker.StopNotebookInstanceInput{
+		NotebookInstanceName: aws.String(id),
+	}
+
+	if _, err := conn.StopNotebookInstance(stopOpts); err != nil {
+		return fmt.Errorf("Error stopping sagemaker notebook instance: %s", err)
+	}
+
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{
+			sagemaker.NotebookInstanceStatusStopping,
+		},
+		Target:  []string{sagemaker.NotebookInstanceStatusStopped},
+		Refresh: sagemakerNotebookInstanceStateRefreshFunc(conn, id),
+		Timeout: 10 * time.Minute,
+	}
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("error waiting for sagemaker notebook instance %q to stop: %s", id, err)
+	}
+
+	return nil
+}
+
+func sagemakerNotebookInstanceStateRefreshFunc(conn *sagemaker.SageMaker, name string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		describeNotebookInput := &sagemaker.DescribeNotebookInstanceInput{
+			NotebookInstanceName: aws.String(name),
+		}
+		notebook, err := conn.DescribeNotebookInstance(describeNotebookInput)
+		if err != nil {
+			if sagemakerErr, ok := err.(awserr.Error); ok && sagemakerErr.Message() == "RecordNotFound" {
```

Comment:
Should this be RecordNotFound or ResourceNotFound to match others?

Reply:
RecordNotFound is the correct usage here